cpp/src/arrow/columnar_format.cc

```diff
@@ -0,0 +1,1034 @@
+template<typename data_type>
+typename std::enable_if<std::is_base_of<BinaryType, data_type>::value, Status>::type
+CopyValue(const typename TypeTraits<data_type>::ArrayType& array, int64_t i,
+          typename TypeTraits<data_type>::BuilderType* builder) {
+  int32_t length;
+  const uint8_t* value = array.GetValue(i, &length);
+  return builder->Append(value, length);
+}
+
+
+#define PRIMITIVE_CASE_LIST \
+  CASE(TimestampType) \
+  CASE(BooleanType) \
+  CASE(Int8Type) \
+  CASE(UInt8Type) \
+  CASE(Int16Type) \
+  CASE(UInt16Type) \
+  CASE(Int32Type) \
+  CASE(UInt32Type) \
+  CASE(Int64Type) \
+  CASE(UInt64Type) \
+  CASE(FloatType) \
+  CASE(DoubleType) \
+  CASE(BinaryType) \
+  CASE(StringType) \
+  CASE(FixedSizeBinaryType) \
+  CASE(Date32Type) \
+  CASE(Date64Type) \
+  CASE(Time32Type) \
+  CASE(Time64Type)
+
+} // anonymous namespace
+
+
+// Shredder implementation
+namespace {
+
+class ShredNode {
+ public:
+  ShredNode(const std::shared_ptr<Field>& field, MemoryPool* pool)
+      : field_(field),
+        rep_levels_builder_(std::make_shared<Int16Builder>(pool)),
+        def_levels_builder_(std::make_shared<Int16Builder>(pool)) {
+  }
+  virtual ~ShredNode() {
+  }
+
+  static Status CreateTree(const std::shared_ptr<Field>& field,
+                           int16_t rep_level, int16_t def_level,
+                           MemoryPool* pool,
+                           std::unique_ptr<ShredNode>* node);
+
+  const std::shared_ptr<Field>& field() const { return field_; }
+  int num_children() const { return static_cast<int>(children_.size()); }
+  ShredNode* child(int i) const { return children_[i].get(); }
+
+  virtual Status Handle(const Array& array, int64_t i, int16_t rep_level) = 0;
+  virtual Status Finish(ColumnMap* colmap) = 0;
+
+  Status FinishRecursive(ColumnMap* colmap) {
+    RETURN_NOT_OK(Finish(colmap));
+    for (int i = 0; i < num_children(); i++) {
+      RETURN_NOT_OK(child(i)->FinishRecursive(colmap));
+    }
+    return Status::OK();
+  }
+
+  // Append given pair of levels recursively to this node and all nodes below
+  Status AppendLevelsRecursive(int16_t rep_level, int16_t def_level) {
+    RETURN_NOT_OK(rep_levels_builder_->Append(rep_level));
+    RETURN_NOT_OK(def_levels_builder_->Append(def_level));
+    for (int i = 0; i < num_children(); i++) {
+      RETURN_NOT_OK(child(i)->AppendLevelsRecursive(rep_level, def_level));
+    }
+    return Status::OK();
+  }
+
+
+  std::shared_ptr<Field> field_;
+  // Repetition level starts with 0 in root element and increases by one in
+  // every list _element_ node.
+  int16_t rep_level_{-1};
+  // Minimal definition level at which this node is not NULL and not empty (if node
+  // is a list).
+  int16_t def_level_{-1};
+  std::shared_ptr<Int16Builder> rep_levels_builder_;
+  std::shared_ptr<Int16Builder> def_levels_builder_;
+  std::vector<std::unique_ptr<ShredNode>> children_;
+};
+
+
+template<typename data_type>
+class PrimitiveShredNode : public ShredNode {
```

Comment:
Two high-level but potentially related comments.  I haven't thought through fully if it is possible but it seems like it could be:
   - Can this be vectorized to calculate a column at a time and then recurse instead of a row at a time?
   - If we can do one, can we expose a callback method so we don't need to keep already computed columns arround (it seems like we might do that here but if we don't you can ignore this).

Reply:
To answer the first point, after having some time to think about it, I don't think this would be an obvious win.